core/src/main/kotlin/fr/sncf/osrd/stdcm/graph/STDCMPathfindingB.kt

```diff
@@ -0,0 +1,289 @@
+package fr.sncf.osrd.stdcm.graph
+
+import fr.sncf.osrd.api.FullInfra
+import fr.sncf.osrd.api.pathfinding.constraints.*
+import fr.sncf.osrd.api.pathfinding.makeHeuristics
+import fr.sncf.osrd.envelope_sim.allowances.utils.AllowanceValue
+import fr.sncf.osrd.graph.*
+import fr.sncf.osrd.reporting.exceptions.ErrorType
+import fr.sncf.osrd.reporting.exceptions.OSRDError
+import fr.sncf.osrd.sim_infra.api.Block
+import fr.sncf.osrd.sim_infra.api.BlockId
+import fr.sncf.osrd.stdcm.STDCMResult
+import fr.sncf.osrd.stdcm.STDCMStep
+import fr.sncf.osrd.stdcm.infra_exploration.initInfraExplorerWithEnvelope
+import fr.sncf.osrd.stdcm.preprocessing.interfaces.BlockAvailabilityInterface
+import fr.sncf.osrd.train.RollingStock
+import fr.sncf.osrd.utils.units.Offset
+import fr.sncf.osrd.utils.units.meters
+import java.time.Duration
+import java.time.Instant
+import java.util.*
+import kotlin.collections.ArrayList
+import kotlin.collections.HashSet
+
+data class EdgeLocation(val edge: STDCMEdge, val offset: Offset<STDCMEdge>)
+
+data class Result(
+    val edges: List<STDCMEdge>, // Full path as edge ranges
+    val waypoints: List<EdgeLocation>
+)
+
+fun findPath(
+    fullInfra: FullInfra,
+    rollingStock: RollingStock,
+    comfort: RollingStock.Comfort?,
+    startTime: Double,
+    steps: List<STDCMStep>,
+    blockAvailability: BlockAvailabilityInterface,
+    timeStep: Double,
+    maxDepartureDelay: Double,
+    maxRunTime: Double,
+    tag: String?,
+    standardAllowance: AllowanceValue?,
+    pathfindingTimeout: Double
+): STDCMResult? {
+    return STDCMPathfindingB(
+            fullInfra,
+            rollingStock,
+            comfort,
+            startTime,
+            steps,
+            blockAvailability,
+            timeStep,
+            maxDepartureDelay,
+            maxRunTime,
+            tag,
+            standardAllowance,
+            pathfindingTimeout
+        )
+        .findPath()
+}
+
+class STDCMPathfindingB(
+    val fullInfra: FullInfra,
+    val rollingStock: RollingStock,
+    val comfort: RollingStock.Comfort?,
+    val startTime: Double,
+    val steps: List<STDCMStep>,
+    val blockAvailability: BlockAvailabilityInterface,
+    val timeStep: Double,
+    val maxDepartureDelay: Double,
+    val maxRunTime: Double,
+    val tag: String?,
+    val standardAllowance: AllowanceValue?,
+    val pathfindingTimeout: Double = 120.0,
+    var graph: STDCMGraph? = null,
```

Comment:
I don't think the graph should be nullable, it could be set during the init. It may have to be moved out of the class declaration though (at the same level as the queue)